src/Router/Middleware/ErrorHandler.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+namespace Jasny\Router\Middleware;
+
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Message\ResponseInterface;
+
+/**
+ * Handle error in following middlewares/app actions
+ */
+class ErrorHandler
+{    
+    /**
+     * Run middleware action
+     *
+     * @param ServerRequestInterface $request
+     * @param ResponseInterface      $response
+     * @param callback               $next
+     * @return ResponseInterface
+     */
+    public function __invoke(ServerRequestInterface $request, ResponseInterface $response, $next = null)
+    {
+        if ($next && !is_callable($next)) {
+            throw new \InvalidArgumentException("'next' should be a callback");            
+        }
+
+        $error = false;
+
+        try {
+            $response = $next ? call_user_func($next, $request, $response) : $response;
+        } catch(\Throwable $e) {
+            $error = true;
+        } catch(\Exception $e) { #This block can be removed when migrating to PHP7, because Throwable represents both Exception and Error
+            $error = true;
+        }
+
+        return $error ? $this->handleError($response) : $response;
+    }
+
+    /**
+     * Handle caught error
+     *
+     * @param ResponseInterface $response
+     * @return ResponseInterface
+     */
+    protected function handleError($response)
+    {
+        $body = $response->getBody();        
+        $body->rewind();
```

Comment:
`rewind()` will not clear the existing data.

```php
$fp = fopen('php://temp', 'r+');
fwrite($fp, "abcdefghijklmnopqrstuvwxyz");
fseek($fp, 0);
fwrite($fp, "1234567890");
fseek($fp, 0);
echo fread($fp, 1024); // outputs "1234567890klmnopqrstuvwxyz"
```

Instead we need to create a new body.

Add a `withStreamFactory($callback)` and `getStreamFactory()` method. The `with..` method should return a clone of `$this` (the ErrorHandler middleware).

If it isn't set, try to see if the `Psr7Middlewares\Middleware` class is loaded and use `Psr7Middlewares\Middleware::getStreamFactory()`.

If that isn't set either, see if it's an instance of `Jasny\HttpMessage\Stream`. In that case you can `clone` the body.

If all fails, throw an exception.